labs/b9.md

```diff
@@ -0,0 +1,178 @@
+﻿---
+title: Lab 9 - Containers and Configuration Management 
+layout: lab
+---
+
+## Containers and Configuration Management    
+
+This exercise is designed to give you some hands-on experience with Docker and Puppet! By the end of this assignment, you should be able to:    
+
+- Create and use a Docker container interactively  and create a Dockerfile, which allows you to declaratively define your containers.
+- Write a basic Puppet manifest and apply the configuration to your VM.
+
+#### Installing Docker    
+
+One of the annoying things about Docker is it’s usually not in `apt` and, if it is, it’s always under a different name. Here’s a link to the .deb for Docker Community Edition you can download and install on your VM: [https://download.docker.com/linux/debian/dists/stretch/pool/stable/amd64/docker-ce\_18.03.0~ce-0~debian\_amd64.deb](https://download.docker.com/linux/debian/dists/stretch/pool/stable/amd64/docker-ce_18.03.0~ce-0~debian_amd64.deb). I trust by this point in the course installing this on your Linux machine should be a breeze :).    
+
+Hint: use `wget` and `dpkg`.    
+
+After installing, I recommend following [these instructions](https://docs.docker.com/engine/installation/linux/linux-postinstall/#manage-docker-as-a-non-root-user) so you can use `docker` as a non-root user. This means you won’t have to type ‘`sudo docker`’ all the time. This is optional but for the rest of this exercise I’m going to assume that you did this. You might see some weird output like:    
+
+>sent invalidate(passwd) request, exiting  
+sent invalidate(group) request, exiting      
+
+But it should work anyway.    
+
+#### Creating your first Docker container    
+
+To verify that you installed things correctly, try running:    
+
+`docker run hello-world`    
+
+You should see some friendly output like so:  
+
+>Unable to find image 'hello-world:latest' locally  
+latest: Pulling from library/hello-world  b04784fba78d: Already exists  
+Digest: sha256:f3b3b28a45160805bb16542c9531888519430e9e6d6ffc09d72261b0d26ff74f
+Status: Downloaded newer image for hello-world:latest  Hello from Docker!    
+
+This message shows that your installation appears to be working correctly. To generate this message, Docker took the following steps:    
+
+1. The Docker client contacted the Docker daemon.  
+2. The Docker daemon pulled the "hello-world" image from the Docker Hub.    
+3. The Docker daemon created a new container from that image which runs the executable that produces the output you are currently reading.    
+ 4. The Docker daemon streamed that output to the Docker client, which sent it to your terminal.    
+
+Some [quick definitions from Docker’s website](https://docs.docker.com/get-started/#a-brief-explanation-of-containers):    
```

Comment:
I think adding some context for this would be useful, instead of just giving the definitions themselves. Maybe this should also go before they run anything? Idk